### PR TITLE
Tool Home Page

### PR DESCRIPTION
**Pull request by ChinemeremChigbo**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Tool-Builder-Form-5a6b5a9e8edb4c13ba8f1e43adeebd0e)

[Deploy Preview](https://deploy-preview-33--modest-payne-4c3f5b.netlify.app/create-a-tool)

## Implementation description
* Admin use the Tool Builder to dynamically generate landing pages for each Tool as well as the Tool Self Check. It primarily consists of a form in which values are provided for elements of Tool landing pages which vary based on each Tool.

## Steps to test
1. Compare to Design to Figma > Builders > Create a Tool
https://www.figma.com/file/GhDd9C0XsyVDgwL5zzvDvO/HUG-S21?node-id=1059%3A16927
2. Ensure that changes to input fields persist when switching between the home and self check pages
3. Ensure that the "Save" Button does not allow you to save an empty document or the last saved document
4. Endure that submit is only allowed when all required fields are filled and that pointing at the greyed out submit button tells you the correct field that needs to be filled out next
5. Ensure that the objects that are printed out when saved/publish match the information you put in
6. Ensure that on the self check question page, you are unable ONLY numbers in the corresponding values AFTER clicking the "123" button
7. Type some information into a multiple choice / multi-select question, change the question type to a short/long answer question and ensure that your original information persists upon switching back. (This will not be the case for the slider as it needs to override the "value" field to keep track of whether the slider should start at 0 or 1)
8. Ensure that adding and moving around questions all over the place produces no errors in the web console.
9. Ensure that the discard clears both the home page and  self check as expected
10. Ensure that the option removal, delete buttons + question button and (Tool Home Page) add link removal buttons work as intended.

## What should reviewers focus on?
* Design
* Functionality

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Review by yashgajaria — CHANGES_REQUESTED** [review]
Tested the functionality and noticed a couple of minor things: 

- If fields are filled in on `Tool Home Page` tab and tabs are changed to `Tool Self Check`, all the fields other than `Title` are reset 
- Under the `Select Recommended Tools` tab, we should only be allowed to select a specific tool once (i.e this image below shouldn't be possible) 
image
- Clarify what the `Add Link` hyperlinks should do under `Related Resources/Stories/External Resources` 
- Try to restrict the `thumbnail` upload field to only image files, I think right now it is still possible to upload other files such as a pdf
- Clarify/add the `publish` button on the `Tool Home Page` and `Tool Self Check`
- Clarify whether `save` should be disabled if fields are empty or how to notice changes in the tool information 
- Under `Tool Self Check` if the `value` is set to `123` it is still possible to input letters in Firefox, however this is not a problem in Chrome
image

**Comment by ChinemeremChigbo** [comment]
> Tested the functionality and noticed a couple of minor things:
> 
> * If fields are filled in on `Tool Home Page` tab and tabs are changed to `Tool Self Check`, all the fields other than `Title` are reset
> * Under the `Select Recommended Tools` tab, we should only be allowed to select a specific tool once (i.e this image below shouldn't be possible)
>   image
> * Clarify what the `Add Link` hyperlinks should do under `Related Resources/Stories/External Resources`
> * Try to restrict the `thumbnail` upload field to only image files, I think right now it is still possible to upload other files such as a pdf
> * Clarify/add the `publish` button on the `Tool Home Page` and `Tool Self Check`
> * Clarify whether `save` should be disabled if fields are empty or how to notice changes in the tool information
> * Under `Tool Self Check` if the `value` is set to `123` it is still possible to input letters in Firefox, however this is not a problem in Chrome
>   image

Should be all done! Feel free to re-review.

File: pages/create-a-tool.tsx [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+                    variant="link"
+                >
+                    Tool Self Check
+                </Button>
+            </Flex>
+            <SimpleGrid columns={1} spacing={0} px={10} py={10}>
+                {page == "home" && (
+                    <>
+                        {toolList.map((item, index) => (
+                            <ToolHomePage
```
Comment (tawnyzhao) [inline pages/create-a-tool.tsx:432]: Add a `key` attribute

File: src/components/toolHomePage/index.tsx [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    ModalContent,
+    ModalHeader,
+    ModalBody,
+    ModalFooter,
+} from "@chakra-ui/react";
+
+const recommendedToolsList = ["Tool 1", "Tool 2", "Tool 3", "Tool 4", "Tool 5"];
+
+//Self check question card component
+export const ToolHomePage: React.FC = ({
```
Comment (tawnyzhao) [inline src/components/toolHomePage/index.tsx:25]: Add types to this component

File: pages/create-a-tool.tsx [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+            if (target != "") {
+                toast({
+                    title: "Invalid file type",
+                    description: "Please upload a jpg, jpeg, bmp, gif, or png",
+                    status: "error",
+                    position: "bottom-left",
+                    duration: 5000,
+                    isClosable: true,
+                });
+                document.getElementById("thumbnail").value = "";
```
Comment (tawnyzhao) [inline pages/create-a-tool.tsx:98]: pass the thumbnail to `ToolHomePage` as a prop and set the value in `Input` line 92 in ToolHomePage

**Review by tawnyzhao — APPROVED** [review]

**Review by tawnyzhao — CHANGES_REQUESTED** [review]
Address comments

File: pages/create-a-tool.tsx (hunk 3 above)
Comment (ChinemeremChigbo) [reply pages/create-a-tool.tsx:98]: Changed to what we discussed.

File: pages/create-a-tool.tsx [hunk 4]
```diff
@@ -0,0 +1,667 @@
...
+            alphanumericInput: true,
+            questionNumber: 1,
+        },
+    ];
+    const [questionList, setQuestionList] = useState(defaultQuestions);
+    const [lastSavedQuestions, setLastSavedQuestions] = useState(
+        JSON.parse(JSON.stringify(defaultQuestions)),
+    );
+    //keeps track of count of questions made so each question has a unique id
+    const [count, setCount] = useState(2);
```
Comment (yashgajaria) [inline pages/create-a-tool.tsx:203]: Quick question why do we use a starting state of 2 for question count?

Comment (ChinemeremChigbo) [reply pages/create-a-tool.tsx:203]: Because that controls the given question ID and the first question ID starts at 1 I believe

Comment (yashgajaria) [reply pages/create-a-tool.tsx:203]: Okay that makes sense- by the questionID are you referring to what's on line 186 as the `defaultQuestions` id?

File: src/components/toolSelfCheck/index.tsx [hunk 5]
```diff
@@ -0,0 +1,873 @@
...
+                                        </InputGroup>
+                                    ))}
+                                    <Button
+                                        maxWidth={"120"}
+                                        variant="ghost"
+                                        onClick={() => {
+                                            onAddOption(questionId, "bottom");
+                                        }}
+                                    >
+                                        +Add Option
```
Comment (yashgajaria) [inline src/components/toolSelfCheck/index.tsx:282]: I think the button to add an option is repeated for every single question type- if you want and can easily do it consider moving it to a top level function so you don't need to repeat yourself

File: src/components/toolSelfCheck/index.tsx [hunk 6]
```diff
@@ -0,0 +1,873 @@
...
+                                                        : "white"
+                                                }
+                                                background={
+                                                    alphanumeric
+                                                        ? "white"
+                                                        : "black"
+                                                }
+                                            >
+                                                123
+                                            </Button>
```
Comment (yashgajaria) [inline src/components/toolSelfCheck/index.tsx:364]: similar to the above comment- this block of code with the alphanumeric buttons is repeated I think 4 times, try moving it to the top level, if not, dw too much about it

**Review by yashgajaria — APPROVED** [review]
A couple of minor comments to be addressed/considered, but overall lgtm. 

One minor error/bug I noticed:
- When on the `home page` if you `add link` under the related resources/stories/external section and then move to the `tool self check` tab and come back to the `home page`, you are unable to delete the links. The `X` doesn't delete the link for any links added from the initial visit to the `home page` tab. 

image

File: pages/create-a-tool.tsx [hunk 7]
```diff
@@ -0,0 +1,667 @@
...
+    useDisclosure,
+    useToast,
+} from "@chakra-ui/react";
+import { SelfCheckQuestionCard, ToolHomePage } from "@components";
+
+//Self Check Questions React functional component
+const Home: React.FC = () => {
+    //Self check tool object
+    const defaultTool = {
+        _id: "50e642d7e4a1ae34207a92a0",
```
Comment (tawnyzhao) [inline pages/create-a-tool.tsx:24]: Remove the `_id` field

File: pages/create-a-tool.tsx [hunk 8]
```diff
@@ -0,0 +1,667 @@
...
+        const tempQuestion = newList[index];
+        newList[index] = newList[index - direction];
+        newList[index - direction] = tempQuestion;
+        setQuestionList(newList);
+        changeQuestionNumbers(newList);
+    };
+
+    const addOneQuestion = (index) => {
+        const newQuestion = {
+            _id: `60e642d7e4a1ae34207a92a${count}`,
```
Comment (tawnyzhao) [inline pages/create-a-tool.tsx:350]: Remove `_id` field, this will be auto generated by the Model on creation

File: pages/create-a-tool.tsx [hunk 9]
```diff
@@ -0,0 +1,667 @@
...
+            }
+        }
+        setStillNeeded(needed);
+        return fieldsFilled;
+    };
+
+    //self check card
+    const defaultQuestions = [
+        {
+            _id: "60e642d7e4a1ae34207a92a0",
```
Comment (tawnyzhao) [inline pages/create-a-tool.tsx:186]: Remove `_id` field

**Review by tawnyzhao — APPROVED** [review]

**Pull request merged**